### ISSUE TITLE
Generic Kotlin data classes cannot be (de)serialized

### ISSUE THREAD

**Issue by svenjacobs** [square/moshi#309]
It seems that Moshi 1.5.0 has a problem with (de)serialization of generic data classes.

```kotlin
data class Test<out T>(val item: T)

data class Test2(val test: Test<String>)
```

The following code

```kotlin
val json = """{test:{item:"hello"}}"""
val moshi = Moshi.Builder()
        .add(KotlinJsonAdapterFactory())
        .build()
val data = moshi.adapter(Test2::class.java).fromJson(json)
```

produces the exception

```
java.lang.IllegalArgumentException: Expected a Class, ParameterizedType, or GenericArrayType, but <null> is of type null
  at com.squareup.moshi.Types.getRawType(Types.java:167)
  at com.squareup.moshi.ClassJsonAdapter$1.createFieldBindings(ClassJsonAdapter.java:83)
  at com.squareup.moshi.ClassJsonAdapter$1.create(ClassJsonAdapter.java:75)
  at com.squareup.moshi.Moshi.adapter(Moshi.java:100)
  at com.squareup.moshi.KotlinJsonAdapterFactory.create(KotlinJsonAdapter.kt:184)
  at com.squareup.moshi.Moshi.adapter(Moshi.java:100)
  at com.squareup.moshi.KotlinJsonAdapterFactory.create(KotlinJsonAdapter.kt:184)
  at com.squareup.moshi.Moshi.adapter(Moshi.java:100)
  at com.squareup.moshi.Moshi.adapter(Moshi.java:62)
  at com.svenjacobs.test.components.main.view.MainActivity.onCreate(MainActivity.kt:24)
```

This was working in Moshi 1.4.0 and before.

**Comment by swankjesse** [comment]
Can fix.

**Comment by fmrsabino** [comment]
Not sure if related but I'm also getting the same issue using Lists. As an example:

```kotlin
data class PaginatedResponse<out T>(val page: Int, val results: List<T>)
```

**Comment by thatJavaNerd** [comment]
Is there a suggested workaround until this issue gets fixed?

**Comment by fmrsabino** [comment]
@thatJavaNerd I reverted to version `1.4.0`

**Comment by cre8ivejp** [comment]
Anyone knows when the PR above fixing this issue will be merged?

Thanks.

**Comment by swankjesse** [comment]
I think it’s too simple. Will try to find time for it.

**Comment by rodly** [comment]
@swankjesse Is this fix being worked on still?

**Comment by swankjesse** [comment]
Yep.

**Comment by mradzinski** [comment]
@swankjesse Any news regarding this? Just wondering if to use Moshi 1.4.0 while this is fixed.

**Comment by ishitatsuyuki** [comment]
Ping, this issue is quite critical.

**Comment by mhf-air** [comment]
Please fix this

**Comment by blackberry2016** [comment]
The same problem.
```java
data class Note constructor(@Json(name = "note_title") var title: String?,
                            @Json(name = "note_star") var star: Long,
                            @Json(name = "note_content") var content: String?)

data class Page<T> constructor(@Json(name = "my_list") var list: List<T>)
```
The following code
```java
@JvmOverloads
fun toJson(obj: Any, type: Type, moshi: Moshi? = null): String {
    val mMoshi = moshi ?: Moshi.Builder().add(KotlinJsonAdapterFactory()).build()
    return mMoshi.adapter<Any>(type).lenient().toJson(obj)
}


val list = listOf(
                Note("note 1", 1024L, "Learn more, sleep more, and eat more-1"),
                Note("note 2", 512L, "Learn more, sleep more, and eat more-2"),
                Note("note 3", 72L, "Learn more, sleep more, and eat more-3"))

toJson(Page(list), Types.newParameterizedType(Page::class.java, Note::class.java))
```
exception
```java
java.lang.IllegalArgumentException: Expected a Class, ParameterizedType, or GenericArrayType, but <null> is of type null

	at com.squareup.moshi.Types.getRawType(Types.java:167)
	at com.squareup.moshi.ClassJsonAdapter$1.createFieldBindings(ClassJsonAdapter.java:83)
	at com.squareup.moshi.ClassJsonAdapter$1.create(ClassJsonAdapter.java:75)
	at com.squareup.moshi.Moshi.adapter(Moshi.java:100)
	at com.squareup.moshi.Moshi.adapter(Moshi.java:58)
	at com.squareup.moshi.CollectionJsonAdapter.newArrayListAdapter(CollectionJsonAdapter.java:52)
	at com.squareup.moshi.CollectionJsonAdapter$1.create(CollectionJsonAdapter.java:36)
	at com.squareup.moshi.Moshi.adapter(Moshi.java:100)
	at com.squareup.moshi.KotlinJsonAdapterFactory.create(KotlinJsonAdapter.kt:184)
	at com.squareup.moshi.Moshi.adapter(Moshi.java:100)
```

**Comment by blackberry2016** [comment]
Is this bug maintained now?

**Comment by blackberry2016** [comment]
Is this bug maintained by now?

**Comment by LouisCAD** [comment]
I guess the issue is not present in kotshi as this fixed issue shows: https://github.com/ansman/kotshi/issues/41

**Comment by blackberry2016** [comment]
For this kind of data class, use java to avoid this bug temporarily

**Comment by mattbdean** [comment]
What I ended up doing was using [AutoValue](https://github.com/google/auto/tree/master/value) + [auto-value-moshi](https://github.com/rharter/auto-value-moshi). You don't need reflection so it's gonna be a lot faster if you need to deserialize a ton of models very quickly like [JRAW](https://github.com/mattbdean/JRAW)

**Comment by autonomousapps** [comment]
I can confirm that the [Kotshi](https://github.com/ansman/kotshi) library solves this problem and, moreover, saves us from having to use moshi-kotlin and, therefore, saves us from kotlin-reflect, which contains >10k methods. Which is an important consideration for an Android project.

**Comment by NightlyNexus** [comment]
fixed by #475 for users of the KotlinJsonAdapterFactory.

**Comment by AllenCoder** [comment]
When to release the latest fix？

**Comment by bradynpoulsen** [comment]
I can verify this has been fixed for `KotlinJsonAdapter` in `1.6.0-RC1` for the following sample.

```kotlin
data class Response<Output : Any>(
    val id: String?,
    val result: Output
)

val adapter: JsonAdapter<Response<List<String>>> = moshi.adapter(
    Types.newParameterizedType(
        Response::class.java,  
        Types.newParameterizedType(
            List::class.java, 
            String::class.java
        )
    )
)

adapter.toJson(Response("foo", listOf("bar", "baz"))) // yields: {"id":"foo","result":["bar","baz"]}
```
